### DISCUSSION TITLE
Adding namespace to root element

### DISCUSSION THREAD

**Discussion by PatrickBauer** [downforcetech/fluidxml.php#6] [Questions]
Is there a way to declare a namespace inside the documents root element?

I'd like to use it like here: https://support.google.com/merchants/answer/160589?hl=en&ref_topic=2473799

`<rss version="2.0" xmlns:g="http://base.google.com/ns/1.0">`

My code looks like this:

```
header("Content-type: text/xml");

$rss = FluidXml('rss');
$rss->namespace('g', 'http://base.google.com/ns/1.0');
$rss->setAttribute('version', '2.0');

$channel = $rss->addChild('channel', true);

$items = $channel->addChild('title', 'Tatem Google Shopping List')
->addChild('link', 'https://www.tatem.de')
->addChild('description', '')
->addChild('items', true);

$items->addChild('item', true)
->addChild('title', 'Testname')
->addChild('link', 'a')
->addChild('description', 'a')
->addChild('g:image_link', 'a')
->addChild('g:price', 'a')
->addChild('g:condition', 'a')
->addChild('g:id', 'a');

echo $rss;
```

**Comment by eolandro** [comment]
I have the same problem. i need to to something like this:

`<?xml version="1.0" encoding="UTF-8"?>`
`<BCE:Balanza xmlns:BCE="www.sat.gob.mx/esquemas/ContabilidadE/1_1/BalanzaComprobacion" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="www.sat.gob.mx/esquemas/ContabilidadE/1_1/BalanzaComprobacion http://www.sat.gob.mx/esquemas/ContabilidadE/1_1/BalanzaComprobacion/BalanzaComprobacion_1_1.xsd"  #Version="1.1">`
`<BCE:Ctas NumCta="100" SaldoIni="0.00" Debe="0.00" Haber="0.00" SaldoFin="0.00"/>`
`<BCE:Ctas NumCta="100.01" SaldoIni="0.00" Debe="0.00" Haber="0.00" SaldoFin="0.00"/>`
`</BCE:Balanza>`

how can i do this?

**Comment by kadet1090** [comment]
I've got same trouble, any ideas? I basically need to parse xmpp xml stream.

**Comment by daniele-orlando** [comment]
Hi guys and thanks for the good questions.

Sorry if this particular situations are not covered by the documentation; I will extend the doc to cover these cases too.

Anyway,
@PatrickBauer, what you need can be achieved relying on some DOM internal APIs. So

``` php
// We have to postpone the root creation.
$rss = fluidxml(null);
$rss->namespace('g', 'http://base.google.com/ns/1.0');

$root = $rss->addChild('rss', true);
// Accessing the node as array returns the DOMNode.
// In this case $root[0] is the first and only associated DOMNode.
$root[0]->setAttributeNS('http://www.w3.org/2000/xmlns/' ,'xmlns:g', 'http://base.google.com/ns/1.0');

// Your root has the namespace registered. Continue as usual.
$channel = $rss->addChild('channel', true);
...
```

@eolandro your case is quite similar.

``` php
$xml = fluidxml(null);
$xml->namespace('BCE', '{BCE_URI}');

$root = $xml->add('BCE:Balanza', true);
$root[0]->setAttributeNS('http://www.w3.org/2000/xmlns/' ,'xmlns:xsi', '{XSI_URI}');

$xml->attr('xsi:schemaLocation', '{SCHEMA_URI}');

// Your root has the namespace registered. Continue as usual.
$xml->addChild('BCE:Ctas', [ 'NumCta' => '100' ]);
...
```

**Comment by daniele-orlando** [comment]
I updated the [Getting Started namespace section](https://github.com/servo-php/fluidxml/wiki/Getting-Started:-8.-Namespaces) to reflect these cases.

**Comment by MisterDetour** [comment]
I'm trying to add a namespace to the root element. I'm going for this:

```
<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:g="http://base.google.com/ns/1.0" version="2.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://www.landsofamerica.com/LandFeed/schemas/LandFeedSchema1.0.xsd">
<channel>
</channel>
```

I'm using the following code:

```
$rss = fluidxml( null );
$root = $rss->addChild( 'rss', true );
$channel = $rss->addChild( 'channel', true );
```

which gives me:

```
<?xml version="1.0" encoding="UTF-8"?>
<rss>
  <channel>
  </channel>
</rss>
```

Is it possible to have both the rss tag and the channel tag at the root level?
